From: Chad Monroe <chad@monroe.io>
Date: Tue, 09 Apr 2024 15:02:01 -0700
Subject: [PATCH] net: bridge: fixups/hacks for SmartOS

SmartOS needs to enslave VLAN devices to a VLAN-aware bridge
for xconnect still.

1. adjust BR_PORT_FLAGS_HW_OFFLOAD to not attempt multicast
   to unicast offloading. the only DSA driver we currently use
   is mt7530 and it does not support it. reduce log spam.
2. wan is not part of br0 but vlan devices might need to be
   added to the bridge for xconnect.

Signed-off-by: Chad Monroe <chad@monroe.io>
---
 net/bridge/br_switchdev.c |    3 +--
 net/bridge/br_vlan.c      |    2 +-
 net/dsa/slave.c           |    3 +++
 3 files changed, 5 insertions(+), 3 deletions(-)

--- a/net/bridge/br_switchdev.c
+++ b/net/bridge/br_switchdev.c
@@ -84,8 +84,7 @@ bool nbp_switchdev_allowed_egress(const
 
 /* Flags that can be offloaded to hardware */
 #define BR_PORT_FLAGS_HW_OFFLOAD (BR_LEARNING | BR_FLOOD | BR_PORT_MAB | \
-				  BR_MCAST_FLOOD | BR_BCAST_FLOOD | BR_PORT_LOCKED | \
-				  BR_HAIRPIN_MODE | BR_ISOLATED | BR_MULTICAST_TO_UNICAST)
+				  BR_MCAST_FLOOD | BR_BCAST_FLOOD | BR_PORT_LOCKED) 
 
 int br_switchdev_set_port_flag(struct net_bridge_port *p,
 			       unsigned long flags,
--- a/net/dsa/slave.c
+++ b/net/dsa/slave.c
@@ -2984,12 +2984,15 @@ dsa_prevent_bridging_8021q_upper(struct
 		return NOTIFY_DONE;
 
 	/* Deny enslaving a VLAN device into a VLAN-aware bridge */
+#if 0
+	/* HACK: allow this for xconnect */
 	if (br_vlan_enabled(br) &&
 	    netif_is_bridge_master(info->upper_dev) && info->linking) {
 		NL_SET_ERR_MSG_MOD(ext_ack,
 				   "Cannot enslave VLAN device into VLAN aware bridge");
 		return notifier_from_errno(-EINVAL);
 	}
+#endif
 
 	return NOTIFY_DONE;
 }
